in peft finetune, only the trainable parameters need to be saved

sywangyi opened the pull request:
to reduce the storage size and also save the time of checkpoint saving while using deepspeed for training

# What does this PR do?

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

Fixes # (issue)


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- text models: @ArthurZucker and @younesbelkada
- vision models: @amyeroberts
- speech models: @sanchit-gandhi
- graph models: @clefourrier

Library:

- flax: @sanchit-gandhi
- generate: @gante
- pipelines: @Narsil
- tensorflow: @gante and @Rocketknight1
- tokenizers: @ArthurZucker
- trainer: @muellerzr and @pacman100

Integrations:

- deepspeed: HF Trainer/Accelerate: @pacman100
- ray/raytune: @richardliaw, @amogkam
- Big Model Inference: @SunMarc
- quantization (bitsandbytes, autogpt): @SunMarc and @younesbelkada

Documentation: @stevhliu and @MKhalusova

HF projects:

- accelerate: [different repo](https://github.com/huggingface/accelerate)
- datasets: [different repo](https://github.com/huggingface/datasets)
- diffusers: [different repo](https://github.com/huggingface/diffusers)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Maintained examples (not research project or legacy):

- Flax: @sanchit-gandhi
- PyTorch: See Models above and tag the person corresponding to the modality of the example.
- TensorFlow: @Rocketknight1

 -->

younesbelkada reviewed (approved):
LGTM, what do you think @pacman100 ?

sywangyi commented:
@younesbelkada I have rebased to main. the code quality check failure is not related with the PR

younesbelkada commented:
@sywangyi can you try to re-run `make fixup` with `pip uninstall black && pip install -U ruff==0.1.5` ?

younesbelkada reviewed (approved):
In principle this looks good! Would like to hear @pacman100 's thoughts before a core maintainer approval

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_27825). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

sywangyi commented:
@pacman100 could you help review the PR? Thanks.

pacman100 reviewed (approved):
Thank you @sywangyi for enabling only the trainable parameters to be saved during intermediate checkpointing when using PEFT with DeepSpeed, LGTM! 🚀

amyeroberts commented on src/transformers/trainer.py:
```diff
@@ -2419,7 +2419,13 @@ def _save_optimizer_and_scheduler(self, output_dir):
+            accept_exclude_frozen_parameters = "exclude_frozen_parameters" in set(
+                inspect.signature(self.model_wrapped.save_checkpoint).parameters.keys()
+            )
+            if accept_exclude_frozen_parameters and is_peft_available() and isinstance(self.model, PeftModel):
```
As this is repeated everywhere, let's abstract out this check to a module level function

```py
def _is_peft_model(model):
    return is_peft_available() and isinstance(self.model, PeftModel)
```


```suggestion
            if accept_exclude_frozen_parameters and _is_peft_model(model):
```

sywangyi replied:
done

amyeroberts reviewed (approved):
Thanks for enabling this!

sywangyi committed:
in peft finetune, only the trainable parameters need to be saved
to reduce the storage size and also save the time of checkpoint saving while using deepspeed for training

Signed-off-by: Wang, Yi <yi.a.wang@intel.com>
